aptos-move/move-examples/veiled_coin/sources/veiled_coin.move

```diff
@@ -0,0 +1,1033 @@
+
+    use aptos_framework::account;
+    use aptos_framework::coin::{Self, Coin};
+    use aptos_framework::event::{Self, EventHandle};
+    use aptos_std::bulletproofs::RangeProof;
+    use aptos_std::bulletproofs;
+
+    //
+    // Errors.
+    //
+
+    /// The range proof system needs to support proofs for any number \in [0, 2^{32})
+    const ERANGE_PROOF_SYSTEM_HAS_INSUFFICIENT_RANGE : u64 = 1;
+
+    /// A range proof failed to verify.
+    const ERANGE_PROOF_VERIFICATION_FAILED : u64 = 2;
+
+    /// Account already has `VeiledCoinStore` registered for `CoinType`
+    const EVEILED_COIN_STORE_ALREADY_PUBLISHED: u64 = 4;
+
+    /// Account hasn't registered `VeiledCoinStore` for `CoinType`
+    const EVEILED_COIN_STORE_NOT_PUBLISHED: u64 = 5;
+
+    /// Not enough coins to complete transaction
+    const EINSUFFICIENT_BALANCE: u64 = 6;
+
+    /// Byte vector failed to deserialize to ciphertexts
+    const EDESERIALIZATION_FAILED: u64 = 7;
+
+    /// Byte vector given for deserialization was the wrong length
+    const EBYTES_WRONG_LENGTH: u64 = 8;
+
+    /// Sigma protocol for withdrawals failed
+    const ESIGMA_PROTOCOL_VERIFY_FAILED: u64 = 9;
+
+    /// Ciphertext has wrong value when unwrapping
+    const ECIPHERTEXT_WRONG_VALUE: u64 = 10;
+
+    /// Value used was larger than vector size in vector cut
+    const EEINDEX_OUT_OF_BOUNDS: u64 = 11;
+
+    //
+    // Constants
+    //
+
+    /// The maximum number of bits used to represent a coin's value.
+    const MAX_BITS_IN_VALUE : u64 = 32;
+
+    /// The domain separation tag (DST) used for the Bulletproofs prover.
+    const VEILED_COIN_DST : vector<u8> = b"AptosVeiledCoinExample";
+    const FIAT_SHAMIR_SIGMA_DST : vector<u8> = b"SigmaFiatShamir";
+
+    //
+    // Core data structures.
+    //
+
+    /// Contains the data necessary to prove that a veiled coin transfer can occur,
+    /// given also a sender and a recipient, and one ciphertext each for the sender
+    /// and recipient which encrypt the amount to be sent using each user's public key.
+    struct VeiledTransferProof<phantom CoinType> has drop {
+        updated_balance_proof: RangeProof,
+        transferred_amount_proof: RangeProof,
+        sigma_proof: SigmaProof<CoinType>,
+    }
+
+    /// Contains the data necessary for a sigma protocol proof used for veiled coin transfers.
+    /// A more detailed description can be found in the documentation of verify_withdrawal_sigma_protocol
+    struct SigmaProof<phantom CoinType> has drop {
+        x1: RistrettoPoint,
+        x2: RistrettoPoint,
+        x3: RistrettoPoint,
+        x4: RistrettoPoint,
+        x5: RistrettoPoint,
+        alpha1: Scalar,
+        alpha2: Scalar,
+        alpha3: Scalar,
+        alpha4: Scalar,
+    }
+
+    /// Given a vector `vec`, removes the last `cut_len` elements of `vec` and
+    /// returns them in order
+    fun cut_vector<T>(vec: &mut vector<T>, cut_len: u64): vector<T> {
```

Comment:
why not `std::vector::trim/trim_reverse`?

Reply:
`trim` was really awfully-named: it's not really a trim. We decided to avoid the confusion and just re-implement what we need. Should be the same gas cost.

Reply:
`std::vector::trim` implements things in the reverse order of `cut_vector` which makes it gas inefficient for this application